### PR TITLE
Adding Fstatfs and statfs as well as tests

### PR DESCRIPTION
**Pull request by TristanB22**

### CODE REVIEW

File: src/safeposix/dispatcher.rs [hunk 1]
```diff
@@ -124,9 +122,18 @@ pub extern "C" fn dispatcher(cageid: u64, callnum: i32, arg1: Arg, arg2: Arg, ar
         LSEEK_SYSCALL => {
             cage.lseek_syscall(unsafe{arg1.dispatch_int}, unsafe{arg2.dispatch_isize}, unsafe{arg3.dispatch_int})
         }
+        XSTAT_SYSCALL => {
```
Comment (jesings) [inline src/safeposix/dispatcher.rs:125]: Ideally we're trying to order the dispatcher by syscall number, which was why XSTAT was up where it was. We might want to reorganize the numbers later, but let's keep things in numeric order

File: src/safeposix/dispatcher.rs [hunk 2]
```diff
@@ -182,15 +180,6 @@ pub extern "C" fn dispatcher(cageid: u64, callnum: i32, arg1: Arg, arg2: Arg, ar
...
         }
-        FCNTL_SYSCALL => {
-            cage.fcntl_syscall(unsafe{arg1.dispatch_int}, unsafe{arg2.dispatch_int}, unsafe{arg3.dispatch_int})
-        }
-        DUP_SYSCALL => {
-            cage.dup_syscall(unsafe{arg1.dispatch_int}, None)
-        }
-        DUP2_SYSCALL => {
-            cage.dup2_syscall(unsafe{arg1.dispatch_int}, unsafe{arg2.dispatch_int})
-        }
```
Comment (jesings) [inline src/safeposix/dispatcher.rs:193]: Why did you get rid of all of these syscalls?

File: src/safeposix/syscalls/fs_calls.rs [hunk 3]
```diff
@@ -475,20 +472,20 @@ impl Cage {
...
+                        },
                         _ => {panic!("A file fd points to a socket or pipe");}
                     }
-                }
+                },
                 Socket(_) => {
                     return syscall_error(Errno::EOPNOTSUPP, "fstat", "we don't support fstat on sockets yet");
-                    }
-                Stream(_) => {self._stat_alt_helper(statbuf, STREAMINODE, &metadata);}
+                },
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:487]: We don't want the commas after the }s here.

File: src/safeposix/syscalls/fs_constants.rs [hunk 4]
```diff
@@ -139,6 +141,28 @@ pub struct StatData {
   pub st_ctim: (u64, u64)
 }
 
+//derive eq attributes for testing whether the structs equal other fsdata structs from stat/fstat
+#[derive(Eq, PartialEq, Default)]
+#[repr(C)]
+pub struct FSData {
+  pub f_type: u32,
+  pub f_bsize: u32,
+  pub f_blocks: u32,
```
Comment (jesings) [inline src/safeposix/syscalls/fs_constants.rs:150]: Testing has shown that all of these are u64 as well

File: src/safeposix/syscalls/fs_constants.rs [hunk 5]
```diff
@@ -139,6 +141,28 @@ pub struct StatData {
...
+  pub f_bavail: u64,
+  //total files in the file system -- should be infinite
+  pub f_files: u64,
+  //free files in the file system -- should be infinite
+  pub f_ffiles: u64,
+  pub f_fsid: u64,
+  //not really a limit for naming, but 254 works
+  pub f_namelen: u32,
+  //arbitrary val for blocksize as well
+  pub f_frsize: u32,
```
Comment (jesings) [inline src/safeposix/syscalls/fs_constants.rs:161]: namelen and frsize should be u64 as well, unless tests on your system yield different results (in which case we'll have to figure something out).

File: src/safeposix/syscalls/fs_constants.rs [hunk 6]
```diff
@@ -139,6 +141,28 @@ pub struct StatData {
...
+  //supposed to be 5 bytes, so let's use null characters -- using 8 bytes in this case
```
Comment (jesings) [inline src/safeposix/syscalls/fs_constants.rs:162]: Get rid of this comment, it's outdated in repy

**Review by jesings — COMMENTED** [review]
Mostly looks good, I do wonder whether some of your errors are because you haven't merged in from main in a little while. If so, that's probably a good thing to do

**Review by TristanB22 — COMMENTED** [review]
I think that I fixed the issues that you were talking about -- please let me know, though.

**Review by rennergade — COMMENTED** [review]
Again the meat of this looks fine, but I'm concerned that there's some overlap with your other PR. Seems like these branches are getting messy in a way that's not ideal.

**Pull request closed without merging**